### PR TITLE
Add Thread Prefix forum list & filtering

### LINKED ISSUE #3508: Thread prefixes - use style in ACP
Same as for usergroups we should use styles for thread prefixes in ACP too.

https://github.com/mybb/mybb/issues/3291#issuecomment-442525541

### LINKED ISSUE #3929: Add Thread Prefix forum list & filtering
List assigned forums for each prefix and allow to filter prefixes by forum in the ACP's _Thread Prefixes_ management.

### PR DESCRIPTION
**Pull request by dvz**
Resolves #3929
Resolves #3508

Prefixes sorted by:
1. _All Forums_
2. multiple forums
3. forum name ([natural sort order](https://en.wikipedia.org/wiki/Natural_sort_order))
3. prefix (natural sort order)

### CODE REVIEW

File: admin/modules/config/thread_prefixes.php [hunk 1]
```diff
@@ -510,14 +510,65 @@ function checkAction(id)
 
 	$table = new Table;
 	$table->construct_header($lang->prefix);
+	$table->construct_header($lang->forums);
 	$table->construct_header($lang->controls, array('class' => 'align_center', 'colspan' => 2));
 
 	$prefixes = build_prefixes();
+
+	foreach($prefixes as &$prefix)
```
Comment (yuliu) [inline admin/modules/config/thread_prefixes.php:518]: `build_prefixes()` could return `false` [ref](https://github.com/mybb/mybb/blob/mybb_1822/inc/functions.php#L3648), so it's better to check if it's an array.

File: admin/modules/config/thread_prefixes.php [hunk 2]
```diff
@@ -510,14 +510,65 @@ function checkAction(id)
...
+	{
+		$prefix['forum_fids'] = explode(',', $prefix['forums']);
+	}
+	unset($prefix);
+
+	$forums = cache_forums();
```
Comment (yuliu) [inline admin/modules/config/thread_prefixes.php:524]: The [`get_forum()`](https://github.com/mybb/mybb/blob/mybb_1822/inc/functions.php#L6267) that wraps details of fetching forums data, IMO, is better than retrieving all forums data. Anyway, using `cache_forums()` is valid.

File: admin/modules/config/thread_prefixes.php [hunk 3]
```diff
@@ -530,7 +581,33 @@ function checkAction(id)
...
+	{
+		return 1;
+	}
+
+	if(count($a['forum_fids']) !== count($b['forum_fids'])) {
+		return count($b['forum_fids']) - count($a['forum_fids']);
+	}
+	elseif(count($a['forum_fids']) === 1)
+	{
+		$forums = cache_forums();
```
Comment (yuliu) [inline admin/modules/config/thread_prefixes.php:607]: Same here. It's worth noting that forum IDs might not exist due to data inconsistency, though rarely happens.

**Review by yuliu — COMMENTED** [review]
Some comments may be helpful.

**Comment by Sama34** [comment]
Seems mostly fine:
image
image

Only issue is when there are no prefixes:
image

**Comment by euantorano** [comment]
This seems like a big usability improvement to me, which is especially useful on forums with lots of prefixes.

**Review by Sama34 — APPROVED** [review]

**Comment by Sama34** [comment]
Is this how the review process works? Haha, anyways seems fine to me now.

**Comment by Eldenroot** [comment]
Can we add something like this alongside with this improvement?
https://github.com/mybb/mybb/issues/3508

It would be reallly coool!

**Review by Sama34 — APPROVED** [review]

**Comment by Sama34** [comment]
image

**Comment by Eldenroot** [comment]
@dvz @Sama34 - there is somwthing wrong, style is not loaded if you use classes in global.css :( is possible to ifx that?

**Comment by dvz** [comment]
There's currently no dedicated CSS source that could be included in the ACP and frontend, but you could take advantage of this to avoid modifying core files: https://github.com/mybb/mybb/blob/mybb_1824/admin/inc/class_page.php#L103-L107

**Comment by Eldenroot** [comment]
Please could you elaborate? Thank you!

I use css classes in global.css right now, so it works for frontend

**Comment by dvz** [comment]
Never mind - there's already a `config.css`. In that case it might be best to create a separate file and add it to ACP's `class_page.php` - e.g.:
```php
echo "	<link rel=\"stylesheet\" href=\"../cache/themes/themeID/thread-prefixes.min.css\" type=\"text/css\" />\n";
```

**Comment by Sama34** [comment]
Basically, you need to add your CSS to the AdminCP as well, not just the front end through the theme system.

**Pull request merged**